x/upgrade/abci.go

```diff
@@ -22,7 +24,20 @@ import (
 // skipUpgradeHeightArray is a set of block heights for which the upgrade must be skipped
 func BeginBlocker(k keeper.Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
 	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
+
 	plan, found := k.GetUpgradePlan(ctx)
+
+	if !DowngradeVerified {
+		DowngradeVerified = true
+		lastAppliedPlan, _ := k.GetLastCompletedUpgrade(ctx)
+		// verify if we are running valid app version
+		if !found || !plan.ShouldExecute(ctx) || (plan.ShouldExecute(ctx) && k.IsSkipHeight(ctx.BlockHeight())) {
+			if lastAppliedPlan != "" && !k.HasHandler(lastAppliedPlan) {
+				panic(fmt.Sprintf("Wrong app version, upgrade handler is missing for %s upgrade plan", lastAppliedPlan))
```

Comment:
Doesn't block merge -- can we get the current app version also printed, or is that not obtainable from state machine?

Reply:
I think it is not possible.

Reply:
There is a available in `version.Version`.